tools/node-modules

```diff
@@ -0,0 +1,223 @@
+            cat >&2 <<EOF
+
+*** node_modules ${sha} doesn't match our package.json
+*** refusing to automatically check out node_modules
+
+Options:
+
+    - tools/node-modules checkout --force     # disable this check
+
+    - tools/node-modules rebuild              # npm install with our package.json
+
+$0: *** aborting
+
+EOF
+            exit 1
+        fi
+    fi
+
+    # we're actually going to do this; let's remove the old one
+    cmd_remove
+
+    # and check out the new one
+    # we need to use the tag name here, unfortunately
+    message UNPACK "node_modules  [sha: ${sha}]"
+    git -c advice.detachedHead=false clone ${quiet} "${CACHE_DIR}" -b "${tag}" node_modules
+}
+
+cmd_rebuild() {
+    for file in npm-shrinkwrap.json package-lock.json yarn.lock; do
+        if [ -e "${top_srcdir}/${file}" ]; then
+            echo "Please delete ${file} from ${top_srcdir} and start again"
+            exit 1
+        fi
+    done
+
+    cmd_remove
+
+    npm install
+
+    cp package.json node_modules/.package.json
+    echo /.cache >> node_modules/.gitignore
+
+    git -C node_modules init -b main
+    git -C node_modules add .
+    git -C node_modules -c gc.autoDetach=false commit -a -m "$(sha256sum package.json)"
+
+    cat <<EOF
+Next steps:
+
+  - tools/node-modules push
+
+  - git add node_modules && git commit
+
+EOF
+}
+
+cmd_push() {
+    local tag="sha-$(git -C node_modules rev-parse HEAD)"
+    git -C node_modules tag -f "${tag}" HEAD
+    git -C node_modules push "${PUSH_REMOTE}" "${tag}"
+}
+
+# called from tools/test-static-code
+cmd_test_static_code() {
+    node_modules_update=$(git log --format=%H -n 1 node_modules)
+    if ! git --no-pager log --stat --exit-code "${node_modules_update}".. package.json; then
+        cat <<EOF
+#
+# *** The above commits modify package.json without updating node_modules
+#
+# Try running:
+#
+#   tools/node-modules rebuild
+#
+EOF
+        exit 1
+    fi
+}
+
+# called from Makefile.am
+cmd_make_package_lock_json() {
+    # Run from make to ensure package-lock.json is up to date
+
+    # package-lock.json is used as the stamp file for all things that use
+    # node_modules (mostly webpack), so this is the main bit of glue that
+    # drives the entire process
+
+    # We try our best not to touch package-lock.json unless it actually changes
+
+    # This isn't going to work for a tarball, but as long as
+    # package-lock.json is already there, and newer than package.json,
+    # we're OK
+    if [ ! -e .git ]; then
+        if [ package-lock.json -nt package.json ]; then
+            exit 0
+        fi
+
+        echo "*** Can't update node modules unless running from git" >&2
+        exit 1
+    fi
+
+    # Otherwise, our main goal is to ensure that the node_modules from
+    # the index is the one that we actually have.
+    local sha="$(get_index_gitlink node_modules)"
+    if [ ! -e node_modules/.git ]; then
```

Comment:
Is this always here? I got the impression it might also be in `.git/modules/node_modules`. Or do we sometimes have both?